README
Epeg version 0.01
=================

Epeg is a perl extension for the The Enlightenment Project's Epeg
library. It must be installed prior to installing this module.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Epeg library 0.9.0.
  http://sourceforge.net/project/showfiles.php?group_id=2&package_id=125546

COPYRIGHT AND LICENCE

Copyright (C) 2005 Michael Curtis

But feel free to go ahead and use it. :)


Changes
Revision history for Perl extension Epeg.

0.06  5/26/2005
    - Added destructor that handles _epeg_close() when the object
      goes out of scope. 

0.05  4/27/2005
    - Modified to catch error codes from epeg_encode() and not 
      attempt to free memory twice.
    - Changed get_data() to return undef on failure.
    - Changed write_file() to return 1 on success undef on failure.
    - Changed set_quality() to normalize the input value.
    - Changed resize() to return undef when attempting to size up
      or scale to the same wxh as the original image.

0.04  4/23/2005
    - Fixed divide by zero error in resize if an image is garbage
      and has zero width or height. 
    - Changed to round up when resizing in MAINTAIN_ASPECT_RATIO
      mode.

0.03  4/17/2005
    - Added test cases
    - Cleaned up POD documentation

0.02  4/15/2005
    - Revised API to handle opening from data reference and opening 
      from a file in new()
    - Added get_comment(), set_comment()
    - Added more test cases

0.01  4/13/2005
    - original version; created by h2xs 1.21 with options
        -n EPEG /usr/local/include/Epeg.h


EPEG.pm
package Image::EPEG;

use 5.006;
use strict;
use warnings;
use Carp;

require Exporter;
require DynaLoader;
use AutoLoader;

our @ISA = qw(Exporter DynaLoader);
our %EXPORT_TAGS = ( 'constants' => [ qw(MAINTAIN_ASPECT_RATIO IGNORE_ASPECT_RATIO) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'constants'} } );
our @EXPORT = qw();
our $VERSION = '0.01';

bootstrap Image::EPEG $VERSION;

use constant MAINTAIN_ASPECT_RATIO => 1;
use constant IGNORE_ASPECT_RATIO => 2;

sub new
{
	my $class = shift;
	my $self = bless { img => undef }, $class;
	
	my $input = shift;
	if( ref $input eq 'SCALAR' )
	{
		# new from data
		$self->{ img } = Image::EPEG::_epeg_memory_open( $$input, length($$input) );
	}
	elsif( $input )
	{
		# new from file
		$self->{ img } = Image::EPEG::_epeg_file_open( $input );
	}

	# return undef on a failed open
	return ref $self->img eq 'Epeg_Image' 
		? $self : undef;
}


sub img		{ return $_[0]->{ img }; }
sub height	{ return $_[0]->{ height }; }
sub width	{ return $_[0]->{ width }; }


sub get_height
{
	my $self = shift;
	$self->_init_size() unless( $self->height );
	return $self->height;
}


sub get_width
{
	my $self = shift;
	$self->_init_size() unless( $self->width );
	return $self->width;
}


sub _init_size
{
	my $self = shift; 
	($self->{ width }, $self->{ height }) =
		Image::EPEG::_epeg_size_get( $self->img );
}


sub set_quality
{
	my $self = shift;
	my $quality = shift;
	Image::EPEG::_epeg_quality_set( $self->img, $quality );
}


sub set_comment
{
	my $self = shift;
	my $comment = shift;
	Image::EPEG::_epeg_comment_set( $self->img, $comment );
}


sub get_comment
{
	my $self = shift;
	return Image::EPEG::_epeg_comment_get( $self->img );
}


sub resize
{
	my $self = shift;
	my $width = shift;
	my $height = shift;
	my $aspect_ratio_mode = shift || IGNORE_ASPECT_RATIO;
	
	# ignore the aspect ratio
	if( $aspect_ratio_mode == IGNORE_ASPECT_RATIO )
	{
		Image::EPEG::_epeg_decode_size_set( $self->img, $width, $height );
		return 1;
	}

	# maintain the aspect ratio
	my ($w, $h) = ($self->get_width(), $self->get_height());
	my ($new_w, $new_h) = (0, 0);

	if( $w * $height > $h * $height )
	{
		$new_w = $width;
		$new_h = $height * $h / $w;
	}
	else
	{
		$new_h = $height;
		$new_w = $width * $w / $h;
	}

	Image::EPEG::_epeg_decode_size_set( $self->img, $new_w, $new_h );
	return 1;
}


sub get_data
{
	my $self = shift;
	my $data = Image::EPEG::_epeg_get_data( $self->img );
	Image::EPEG::_epeg_close( $self->img );
	return $data; 
}


sub write_file
{
	my $self = shift;
	my $path = shift;
	Image::EPEG::_epeg_write_file( $self->img, $path );
	Image::EPEG::_epeg_close( $self->img );
	return 1;
}


1;

__END__


=head1 NAME

EPEG - Perl extension for EPEG

=head1 SYNOPSIS

  use Image::EPEG qw(:constants);
  my $epg = new Image::EPEG( "test.jpg" );
  $epg->resize( 150, 150, MAINTAIN_ASPECT_RATIO );
  $epg->write_file( "test_resized.jpg" );

=head1 DESCRIPTION

Perl wrapper to the alarmingly fast jpeg manipulation library "Epeg".

=head2 Methods

=over 4

=item * new( [FILENAME|DATA REFERENCE] )

=item * get_height()

=item * get_width()

=item * set_quality( [0-100] )

=item * resize( WIDTH, HEIGHT, [Aspect Ratio Mode] )

=item * write_file( FILENAME )

=item * get_data()

=back

=head1 AUTHOR

Michael Curtis E<lt>mcurtis@yahoo-inc.com<gt>

=head1 SEE ALSO

L<http://gatekeeper.dec.com/pub/BSD/NetBSD/NetBSD-current/pkgsrc/graphics/epeg/README.html>

=cut


Epeg.pm
package Image::Epeg;

use 5.006;
use strict;
use warnings;
use Carp;

require Exporter;
require DynaLoader;
use AutoLoader;

our @ISA = qw(Exporter DynaLoader);
our %EXPORT_TAGS = ( 'constants' => [ qw(MAINTAIN_ASPECT_RATIO IGNORE_ASPECT_RATIO) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'constants'} } );
our @EXPORT = qw();
our $VERSION = '0.06';

bootstrap Image::Epeg $VERSION;

use constant MAINTAIN_ASPECT_RATIO => 1;
use constant IGNORE_ASPECT_RATIO => 2;

sub new
{
	my $class = shift;
	my $self = bless { img => undef }, $class;
	
	my $input = shift;
	if( ref $input eq 'SCALAR' )
	{
		# new from data
		$self->{ img } = Image::Epeg::_epeg_memory_open( $$input, length($$input) );
	}
	elsif( $input )
	{
		# new from file
		$self->{ img } = Image::Epeg::_epeg_file_open( $input );
	}

	# return undef on a failed open
	return ref $self->img eq 'Epeg_Image' 
		? $self : undef;
}


sub DESTROY
{
	my $self = shift;
	Image::Epeg::_epeg_close( $self->img )
		if( ref $self->img eq 'Epeg_Image' );
}


sub img		{ return $_[0]->{ img }; }
sub height	{ return $_[0]->{ height }; }
sub width	{ return $_[0]->{ width }; }


sub get_height
{
	my $self = shift;
	$self->_init_size() unless( $self->height );
	return $self->height;
}


sub get_width
{
	my $self = shift;
	$self->_init_size() unless( $self->width );
	return $self->width;
}


sub _init_size
{
	my $self = shift; 
	($self->{ width }, $self->{ height }) =
		Image::Epeg::_epeg_size_get( $self->img );
}


sub set_quality
{
	my $self = shift;
	my $quality = shift;
	$quality = 100 if( $quality < 0 || $quality > 100 );
	Image::Epeg::_epeg_quality_set( $self->img, $quality );
}


sub set_comment
{
	my $self = shift;
	my $comment = shift;
	Image::Epeg::_epeg_comment_set( $self->img, $comment );
}


sub get_comment
{
	my $self = shift;
	return Image::Epeg::_epeg_comment_get( $self->img );
}


sub resize
{
	my $self = shift;
	my $bw = shift;		# boundary width
	my $bh = shift;		# boundary height
	my $aspect_ratio_mode = shift || IGNORE_ASPECT_RATIO;
	
	# make sure we can resize to this wxh
	my ($w, $h) = ($self->get_width(), $self->get_height());
	if( ($w <= 0 || $h <= 0) ||
		($bw > $w && $bh > $h) )
	{
		return undef;
	}

	# ignore the aspect ratio
	if( $aspect_ratio_mode == IGNORE_ASPECT_RATIO )
	{
		Image::Epeg::_epeg_decode_size_set( $self->img, $bw, $bh );
		return 1;
	}

	# maintain the aspect ratio
	my ($new_w, $new_h) = (0, 0);
	if( $w * $bh < $bw * $h )
	{
		$new_w = int($w * $bh/$h + .5);
		$new_h = $bh;
	}
	else
	{
		$new_w = $bw;
		$new_h = int($h * $bw/$w + .5);
	}

	Image::Epeg::_epeg_decode_size_set( $self->img, $new_w, $new_h );
	return 1;
}


sub get_data
{
	my $self = shift;
	my $data = Image::Epeg::_epeg_get_data( $self->img );
	return defined $data ? $data : undef;
}


sub write_file
{
	my $self = shift;
	my $path = shift;
	my $rv = Image::Epeg::_epeg_write_file( $self->img, $path );
	return $rv ? 1 : undef;
}


1;

__END__


=head1 NAME

Epeg - Thumbnail jpegs at lightning speed

=head1 SYNOPSIS

  use Image::Epeg qw(:constants);
  my $epg = new Image::Epeg( "test.jpg" );
  $epg->resize( 150, 150, MAINTAIN_ASPECT_RATIO );
  $epg->write_file( "test_resized.jpg" );

=head1 DESCRIPTION

Perl wrapper to the ultra-fast jpeg manipulation library "Epeg". This library can be used to thumbnail (resize down) jpegs, set comments and quality. NOTE: The resize() method *must* be called with valid arguments or get_data() and write_file() will fail.

=head2 Methods

=over 4

=item * new( [filname|data ref] )

=item * get_height()

=item * get_width()

=item * set_quality( [0-100] )

=item * set_comment( [comment] )

=item * get_comment()

=item * resize( [width], [height], [Aspect Ratio Mode] )

The resize() method can only be used to downsize images. If neither the width or height specified is less than the source image it will return undef.

=item * write_file( [filename] )

=item * get_data()

=back

=head1 AUTHOR

Michael Curtis E<lt>mike@beatbot.comE<gt>

=head1 SEE ALSO

L<http://gatekeeper.dec.com/pub/BSD/NetBSD/NetBSD-current/pkgsrc/graphics/epeg/README.html>

=cut
